File: 2020/25/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	input := strings.Split(string(data), "\n")
	fmt.Println(part1(input))
}

func parseInput(input []string) (int, int) {
	cardPublicKey, _ := strconv.Atoi(input[0])
	doorPublicKey, _ := strconv.Atoi(input[1])
	return cardPublicKey, doorPublicKey
}

func part1(input []string) int {
	cardPublicKey, _ := strconv.Atoi(input[0])
	doorPublicKey, _ := strconv.Atoi(input[1])
	subjectNumber := 7
	loops := 0
	value := 1
	for value != cardPublicKey {
		loops++
		value *= subjectNumber
		value %= 20201227
	}
	cardLoops := loops
	// loops = 0
	// value = 1
	// for value != doorPublicKey {
	// 	loops++
	// 	value *= subjectNumber
	// 	value %= 20201227
	// }
	// doorLoops := loops
	value = 1
	for i := 0; i < cardLoops; i++ {
		value *= doorPublicKey
		value %= 20201227
	}
	// value = 1
	// for i := 0; i < doorLoops; i++ {
	// 	value *= cardPublicKey
	// 	value %= 20201227
	// }
	return value
}

File: 2020/25/solution.ts
import * as fs from "fs";
import * as path from "path";

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n")
  .map(Number);

const cardPublicKey = input[0];
const doorPublicKey = input[1];

let subjNumber = 7;

function transform(value: number, subjNumber: number): number {
  return (value * subjNumber) % 20201227;
}

function decrypt(subjNumber: number, publicKey: number): number {
  let loops = 0;
  let value = 1;
  while (value != publicKey) {
    loops++;
    value = transform(value, subjNumber);
  }
  return loops;
}

function encrypt(subjNumber: number, loops: number): number {
  let value = 1;
  for (let i = 0; i < loops; i++) {
    value = transform(value, subjNumber);
  }
  return value;
}

const cardLoops = decrypt(subjNumber, cardPublicKey);
// const doorLoops = decrypt(subjNumber, doorPublicKey);

// console.log(encrypt(cardPublicKey, doorLoops));
console.log(encrypt(doorPublicKey, cardLoops));
